Support mx master

javitonino opened the pull request:
Adds support for the MX Master (#208), supporting DPI adjusting and the Smart Shift feature. Some details:
- Adds the ability to R/W choices to features (HID 2.0+)
- Smart Shift supports any value from 0x01 to 0xff, I have limited the choices to what the Logitech software allows.

rm225 commented:
Kinda new at this but I couldn't get the DPI changing options to show up with my MX master, it's my first time installing without my distro's package manager so I could have done something wrong.

javitonino commented:
@rm225 You probably downloaded Solaar's main repository. This is not yet merged, so you will have to download from my branch. Basically:
```
git clone https://github.com/javitonino/Solaar.git
cd Solaar
git checkout support-mx-master
```
and assuming you have the dependencies installed, you can run it `bin/solaar` (ensure that you have closed any old version). `bin/solaar config master` (or the GUI) should display the available options (DPI and Smart Shift).

rm225 commented:
Ok, that appears to have worked, I tried using the download zip option previously while browsing files on your fork of the solar repo, must have been on the wrong branch or something, either way, definitely appears to be working now. Hope this gets merged.

shizeon commented:
Have been eagerly awaiting DPI support for MX Master. I tried to run the support-mx-master from your branch but get a invalid syntax error in descriptors.py. I'm not python savy enough to determine why it isn't liking this line..

$ bin/solaar config master
solaar: error: invalid syntax (descriptors.py, line 95)

`_MX_MASTER_SMART_SHIFT_CHOICES = _NamedInts(**{str(a):a for a in range(4,50,4)}, **{'off': 255})`

Any idea? I'm able to run bin/solaar directly out the master branch without errors.

Thank you.

Lekensteyn commented:
Please try the features branch at https://github.com/Lekensteyn/Solaar/tree/features, I have recently updated support for DPI adjustment. This branch should also support future devices with similar functionality, not just the MX Master.


@javitonino The first field is "sensorIndex" which happens to be 0 in this case, but not always. In the features branch I propose an alternative method where the written value is just shifted to account for this byte. (An alternative is a proxy function that performs the required transformation, but that approach is more complex than currently needed.)

shizeon commented:
Hello Lekensteyn,
 I gave your branch a try. From command line:

```
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master 
Wireless Mouse MX Master (MX Master) [4041:F80B37B3]

# dpi
#   possible values: one of [ 1, 16384, 37088, 51206 ], or higher/lower/highest/max/lowest/min
solaar: error: ChoicesValidator: failed to validate read value 3E8
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi higher
solaar: error: coercing to Unicode: need string or buffer, int found
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi lowest
solaar: error: {'params': '\x00\x01\x00', 'request': 2365L, 'number': 1, 'error': 2}
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 16384
solaar: error: {'params': '@\x00\x00', 'request': 2361L, 'number': 1, 'error': 2}
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 37088
solaar: error: {'params': '\x90\xe0\x00', 'request': 2366L, 'number': 1, 'error': 2}
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 51206
solaar: error: {'params': '\xc8\x06\x00', 'request': 2360L, 'number': 1, 'error': 2}
```

And in the GUI is shows there is 1 paired device, but it doesn't list the actual device..
https://dl.dropboxusercontent.com/u/180604/solaar/lekensteyn-solaar-gui

I wonder if mine is somehow different hardware.

```
Unifying Receiver
  Device path  : /dev/hidraw3
  USB id       : 046d:c52b
  Serial       : 247991A2
    Firmware   : 12.03.B0025
    Bootloader : 02.15
    Other      : AA.AA
  Has 1 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)
  Device activity counters: 1=157

  1: Wireless Mouse MX Master
     Codename     : MX Master
     Kind         : mouse
     Wireless PID : 4041
     Protocol     : HID++ 4.5
     Polling rate : 8 ms (125Hz)
     Serial number: F80B37B3
        Bootloader: BOT 18.01.B0014
          Firmware: MPM 11.02.B0014
          Firmware: MPM 11.02.B0014
             Other: 
     The power switch is located on the base.
     Supports 29 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
         3: DEVICE NAME            {0005}   
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: RESET                  {0020}   
         6: BATTERY STATUS         {1000}   
         7: CHANGE HOST            {1814}   
         8: REPROG CONTROLS V4     {1B04}   
         9: ADJUSTABLE DPI         {2201}   
        10: VERTICAL SCROLLING     {2100}   
        11: SMART SHIFT            {2110}   
        12: HIRES WHEEL            {2121}   
        13: GESTURE 2              {6501}   
        14: DFUCONTROL 2           {00C1}   
        15: unknown:1813           {1813}   internal, hidden
        16: unknown:1830           {1830}   internal, hidden
        17: unknown:1890           {1890}   internal, hidden
        18: unknown:18A1           {18A1}   internal, hidden
        19: unknown:18C0           {18C0}   internal, hidden
        20: unknown:1DF3           {1DF3}   internal, hidden
        21: unknown:1E00           {1E00}   hidden
        22: unknown:1EB0           {1EB0}   internal, hidden
        23: unknown:1803           {1803}   internal, hidden
        24: unknown:1861           {1861}   internal, hidden
        25: unknown:9000           {9000}   internal, hidden
        26: unknown:9200           {9200}   internal, hidden
        27: unknown:9240           {9240}   internal, hidden
        28: unknown:1805           {1805}   internal, hidden
     Battery: 50%, discharging.


```

Lekensteyn commented:
Hey @shizeon, thank you for your feedback, please pull my branch and try again. Made small mistake in the parameters.

shizeon commented:
Thanks for looking at this Lekenstyn. I'm still getting a similar error. I'm not seeing the "coercing to Unicode: need string or buffer, int found" error though, but DPI still fails to set.

Here's some sample cli attempts and associated errors:

``
shizeon@asustower:~/Documents/src/Solaar$ git pull
remote: Counting objects: 11, done.
remote: Compressing objects: 100% (11/11), done.
remote: Total 11 (delta 2), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (11/11), done.
From https://github.com/Lekensteyn/Solaar
   9c768d6..dd27559  features   -> origin/features
Updating 9c768d6..dd27559
Fast-forward
 lib/logitech_receiver/settings_templates.py | 8 +++-----
 lib/solaar/cli/config.py                    | 2 +-
 2 files changed, 4 insertions(+), 6 deletions(-)
shizeon@asustower:~/Documents/src/Solaar$ git checkout features
Already on 'features'
Your branch is up-to-date with 'origin/features'.
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config 1 dpi 51206
solaar: error: {'params': '\x00\xc8\x06', 'request': 2361L, 'number': 1, 'error': 2}
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config 1 dpi highest
solaar: error: {'params': '\x00\xc8\x06', 'request': 2362L, 'number': 1, 'error': 2}
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi highest
solaar: error: {'params': '\x00\xc8\x06', 'request': 2365L, 'number': 1, 'error': 2}
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi lower
solaar: error: {'params': '\x00\x90\xe0', 'request': 2363L, 'number': 1, 'error': 2}
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi min
solaar: error: {'params': '\x00\x00\x01', 'request': 2365L, 'number': 1, 'error': 2}
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 37088
solaar: error: {'params': '\x00\x90\xe0', 'request': 2367L, 'number': 1, 'error': 2}
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config 
usage: solaar config [-h] device [setting] [value]
solaar config: error: too few arguments
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config 
usage: solaar config [-h] device [setting] [value]
solaar config: error: too few arguments
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master
Wireless Mouse MX Master (MX Master) [4041:F80B37B3]

# dpi
#   possible values: one of [ 1, 16384, 37088, 51206 ], or higher/lower/highest/max/lowest/min
solaar: error: ChoicesValidator: failed to validate read value 3E8
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config 1
Wireless Mouse MX Master (MX Master) [4041:F80B37B3]

# dpi
#   possible values: one of [ 1, 16384, 37088, 51206 ], or higher/lower/highest/max/lowest/min
solaar: error: ChoicesValidator: failed to validate read value 3E8
shizeon@asustower:~/Documents/src/Solaar$

Lekensteyn commented:
@shizeon Can you try it again? I had to ignore a field.

shizeon commented:
@Lekensteyn thanks. Pulled and now a different error.

```
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master
solaar: error: unpack requires a string argument of length 15
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 51206
solaar: error: unpack requires a string argument of length 15
```

Lekensteyn commented:
Can you add this debugging statement before `for val in _unpack` and show the output?

```py
print("reply = %r" % reply)
```

shizeon commented:
@Lekensteyn as requested:

```
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 51206
reply= '\x00\x01\x90\xe0\xc8\x06@\x00\x00\x00\x00\x00\x00\x00\x00\x00'
solaar: error: unpack requires a string argument of length 15

```

Lekensteyn commented:
Now it should work :-) *fingers crossed*

SlySven commented:
So which one of those characters is the extra one - too many NULLs on the end? :stuck_out_tongue_winking_eye:

_A passing innocent by-stander_ (who **finally** found his missing Nano receiver after buying a **new** Unifying one - and a second-hand cordless keyboard without a receiver at a car-boot sale)! :grin:

shizeon commented:
@Lekensteyn We are getting closer :) Now, DPI list has changed. However, it doesn't take the arguments:

```
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 
solaar: error: Invalid DPI list range: [400, 1600]
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 400
solaar: error: Invalid DPI list range: [400, 1600]
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 1600
solaar: error: Invalid DPI list range: [400, 1600]
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master
solaar: error: Invalid DPI list range: [400, 1600]
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master 400
solaar: error: Invalid DPI list range: [400, 1600]
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master DPI 1600
solaar: error: Invalid DPI list range: [400, 1600]
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi 51206
solaar: error: Invalid DPI list range: [400, 1600]

```

Lekensteyn commented:
@shizeon Please try again, the check was a bit wrong.

@SlySven Nah, the protocol just happens to be defined to use 20 byte "frames" (if you can call them like that), but the protocol payload is defined in a different way (the documentation says that it ends with "00 00", but that would require 21 bytes...)

shizeon commented:
@Lekensteyn That did the trick. Command line switching is working and it is also showing up in the GUI with functional DPI shifting! Thank you.
```
shizeon@asustower:~/Documents/src/Solaar$ bin/solaar config master dpi max
# dpi
dpi = NamedInt(1600, '1600')

```

I run this on Ubuntu 15.10.  There is one issue (completely unrelated to mx master and DPI) when firing up the GUI. Just an FYI.. I had to comment out line 735 in  lib/solaar/ui/window.py . It was dying with a 

```
    Gtk.Window.set_default_icon_from_file(_icons.icon_file(NAME.lower()))
TypeError: Argument 0 does not allow None as a value

```

Might be Ubuntu specific though. Once that line was removed it fires up prefectly. Briefly has no appindicator icon, but then switches to the battery one okay.

sudoeric commented:
@Lekensteyn I'm using your features branch for the MX Master support but I'm noticing that setting a DPI greater than 1000 will revert it back to 1000. I can not reproduce this on demand and I can't tell you how long it takes for it to revert down to 1000, but it's not immediate, but happens overnight.

javitonino commented:
@sudoeric Yes, I can also reproduce this, also with my branch. There is some kind of settings memory in the MX master that resets when the mouse goes into power-saving mode (~15 min without moving).

I am investigating the protocol to see if I can find which commands make the settings persist and submit a patch.

Lekensteyn commented:
@javitonino There are notifications from the receiver when a device disconnects due to inactivity, it is message "0x42" (Connection Status Notification) and can be found in the HID++ 1.0 spec. I think Solaar handles this, but have not recently checked this.

Btw, the DPI support is merged to master. The Smart Shift feature is still missing. I'll ask if I can forward some specifications to you if you are interested?

arigit commented:
@Lekensteyn wondering what is blocking smartshift support from being merged? 
Noticing this branch has conflicts - looking forward to be able to use it to add smartshift sensitivity support. 
If it helps - I have a windows7 environment with the latest proprietary driver and am able to capture USB traces

Phil11W commented:
Is there any branch that has Smart Shift working?

Lekensteyn commented:
@arigit I have protocol specifications for the feature, so a capture file should not be needed. The problem now is how to expose the options in a UI to the user (SetPoint has a fancy UI for that with sliders). If you (or @javitonino) are interested, the specs are https://lekensteyn.nl/files/logitech/x2110_smartshift.html (especially the "Suggested host software behavior")

@Phil11W I believe that this branch (with conflicts) has SmartShift working, but it cannot be merged due to conflicts.

arigit commented:
Took a look at the OEM "Options" SW for windows and it basically gives the same as proposed in lekensteyn's page.
I understand a GUI is key for Solaar but wanted to suggest: since the protocol is well documented would it be possible to start by providing command line support first: a CLI switch to enable/disable smartswitch, and parameter to set the sensitivity (takes a byte value [0-255]),
With this at least we can ensure the functionality works, and expose it. GUI would follow.

0x00 = Do not change
0x01-0xFE = Disengage when wheel rotation exceeds N/4 turns per second
0xFF = Ratchet always engaged

javitonino commented:
Hi! Sorry for disappearing, it has been a busy couple of months. I have updated the branch to be able to merge against current master.

About latest discussion:
- This branch works with the UI, with a set value of choices (equivalent to Windows software) in a combo. Not great, but does the job.
- I have updated the branch with the specs provided by @Lekensteyn (thanks!) so it now saves the default value, so the setting is unchanged after restarting the mouse. This is an UGLY hack, but works for now, I'll see if I can improve this reasonably.

javitonino commented:
Just pushed some more changes to make the UI more similar to the Logitech drivers. It split the settings in two controls, one to set the wheel mode (free-spinning, smart-shift or ratchet) and another to adjust smart-shift sensitivity.

It has the problem that changing the wheel mode to ratchet overwrites the sensitivity setting (as ratchet is implemented as smart-shift with maximum disengage value).

Phil11W commented:
None of the smart-shift settings seem to work for me. They don't change. Neither in the GUI nor CLI. DPI settings work fine.

javitonino commented:
Uhm, it works fine for me with:
```
solaar config master wheel-mode freespin
```
```
solaar config master wheel-mode ratchet
```
```
solaar config master wheel-mode smart
solaar config master smart-shift 24
```

@Phil11W If you read the setting after changing it, do you get the same value you set?

Lekensteyn commented on lib/logitech_receiver/settings.py:
```diff
@@ -356,3 +357,36 @@ def prepare_write(self, new_value, current_value=None):
+		self.max_value = max_value
+		self.needs_current_value = False
+
+		self._bytes_count = math.ceil(math.log(max_value + 1, 256))
```
Apparently Python has a `.bit_length()` function, see also `ChoicesValidator.__init__` (maybe we should add a helper function to `common.py`?)

javitonino replied:
Python 2.7+/3.1+ only. I wasn't sure of the minimum version for Solaar and didn't want to bump requirements, but seeing that it looks like it is already at 2.7/3.2+, so changing.

Lekensteyn commented:
Oops, I am reviewing older commits... when you are ready for submission, can you rebase your branch on the current master such that the changes are more logical? There is some cruft like offsets in choices (should not be needed).

javitonino commented:
@Lekensteyn Sure thing. I worked in top of the old branch and for sure it is going to be hard to track changes.

Before preparing a submission version, I would like an opinion on the controls split. I see the following possible solutions for the CLI/GUI.
 - One setting, numeric. From 0 (freespin) to 255(ratchet) with smart-shift sensitivity all the intermediate values. Similar to this branch at c12c18a, but using a numeric value instead of some pre-defined choices.
 - Two settings, one choice (freespin/smart-shift/ratchet) and one numeric for sensitivity. This branch at 93a868f. Problem is that after changing from ratchet to smart-shift the sensitivity is not remembered and is hard to handle (specially in the CLI, as two commands are needed to change from ratchet to smart-shift)
 - Two settings, one toggle (freespin/smart-shift) and one numeric for sensitivity. Similar to the previous one, but the ratchet mode is less explicit, instead of a choice, implies setting smart shift sensitivity to the extreme. Easier to handle than the previous and maybe could be solved with the setting description.

After trying the second one (latest commit in the branch) I am more convinced that the first idea (single control) fits better within Solaar architecture. It is perhaps a bit unclear for the user, as it has to use the min/max values for freespin/racthet modes, but I think that kind of makes sense (minimum sensitivity = always ratchet, maximum sensitivity = always freespin).

Anyway, I would like to hear some opinions before reaching a decision and cleaning up the code a bit.

Phil11W commented:
@javitonino  No, the values don't change for anything wheel related (dpi works, though). How can I help debug?

> solaar config master wheel-mode freespin
# Wheel mode (Smart Shift)
wheel-mode = NamedInt(1, 'smart')

> solaar config master wheel-mode ratchet
# Wheel mode (Smart Shift)
wheel-mode = NamedInt(1, 'smart')

> solaar config master smart-shift 24
# Smart Shift
smart-shift = 48

> solaar config master dpi
# dpi
#   possible values: one of [ 400, 600, 800, 1000, 1200, 1400, 1600 ], or higher/lower/highest/max/lowest/min
dpi = NamedInt(1000, '1000')

> solaar config master dpi 1600
# dpi
dpi = NamedInt(1600, '1600')

javitonino commented:
Please @Phil11W, contact me directly on email to avoid spamming in this thread too much. You can find my email in the git commits, e.g: by running `git log`.

arigit commented:
Tested Javitonino's branch: https://github.com/javitonino/Solaar/tree/support-mx-master
Overall it's awesome: now SmartShift is finally usable in Linux!
Issues I noticed:
 - The smartshift sensitivity slider behaves in a strange way: the slider can't be 'dragged' with the mouse: any attempt to do that (click/hold + drag slider to left or right) ends up in the value decreasing/increasing by just "1"; this is on Fedora 23 (gnome 3.18.2)
    The only way to change the sensitivity is to click directly in the desired position in the slider line
    (normal behavior would be: click & hold allows the slider to be positioned anywhere within the range bar)

- Usable smartshift sensitivity values seems to be 1 to 40 or so, anything above that and Smartmode behaves like in ratchet mode; this the current problem in the slider (see above), setting low sensitivities is very difficult. Wondering if in the Windows config app the slider they show only allows setting of a limited range (say 1-32)

javitonino commented:
@arigit. I am aware of the first problem, still trying to find a solution to it, as Gtk does not appear to have an "end-of-dragging" event. About the second one, yes, Windows drivers let you set values from 4 to 50, similar to what I have in the first version of this branch. I may go back to only allowing certain values as it is very hard to spin the wheel fast enough to make the smart shif trigger for large values.

arigit commented:
@javitonino - comparing the OEM tool with solaar/your branch, realized that the OEM supports setting the thumbwheel sensitivity in 8 steps, I found it practical since the thumbwheel rotation changes tabs in firefox and the default sensitivity causes inadvertent tab switches every now and then (hence in windows I lowered). It would be cool to have this too in solaar :)

javitonino commented:
@arigit I'll probably take a look at other functionality after smart shift, but I started with this as it was what I needed. I just updated the branch with UI changes which should make this easier to use. Please give it a try and tell me if the slider feels better now.

arigit commented:
@javitonino tested the updated branch, works beautifully and I really like the new approach better, usability issues are solved. Only comment: when the mouse is hovering on the slidebar, it is intuitive to try to change it using the Wheel (that's the standard gnome-shell behavior for slidebars as can be seen in the Shell's sound volume slider) - in addition to click & drag. It would be nice to enable this behavior in the new slidebar as well if no bad interaction happens (I can guess you found some issue with it hence had to disable it; one work-around could be to insert a ~3-second delay between the new slidebar change event and the "Send command" event (in case the slidebar keeps changing)).

javitonino commented:
@arigit Actually, it's just disabled by default and I forgot to enable it. Should be working now. I have a delay (0,5s) already in place, as it was also needed for the dragging case.

hobarrera commented:
I've noticed the original Logitech app supports natural scrolling. I've always had to add a hack for this (that doesn't work on all apps).  
Is it possible to include this as well? Or is it done via software by Logitech's app too.

arigit commented:
@javitonino tried your latest current branch and the slider works great now. 

There is a caveat I found; if when loading solaar the mouse was in freespin mode, a "warning" triangle shows next to the sliderbar saying "read/write operation failed" and the slider is greyed out - however DPI can be adjusted with no problems; only the smartshift is affected. Solaar gets stuck in this error condition until it is a) solaar is restarted with the mouse previously set to ratchet mode (with the mouse button next to the wheel), or b) in solaar interface, click on "Unifying receiver", then set ratchet mode in the mouse, then click on MX Master.

This is very close to fully baked now - hope it can be merged to master soon :)

lexelby commented:
Just tried this and it worked flawlessly.  Install was super-easy and DPI/smartshift settings work perfectly.  I love this mouse and I love this branch :D

lexelby commented:
Logitech take note: if this branch hadn't existed I would not have bought this mouse.

javitonino commented:
@hobarrera I'll take a look at it, in any case, work for another PR
@arigit Should be fixed now. Thanks for all the exhaustive testing. Really appreciated!

arigit commented:
@javitonino thanks for this! been using it for a little while and it works perfectly - no issues. Hoping this PR can be merged Master! (although I would love to see some support for the thumbwheel sensitivity as well :)

javitonino commented:
As the branch appears to be stable now, I have rebased the code on master and reorganized the commits so it is easier to read.

@Lekensteyn can you review it?

Lekensteyn commented:
@javitonino Thanks, I'll review it after two weeks, have to prepare for some exams atm. Hope that is OK :)

javitonino commented:
@Lekensteyn Of course, no problem. Good luck with those exams!

douglaswth commented:
In Logitech Options on Windows I am able to switch the functionality of the wheel button and the top middle button so the wheel button switches the wheel scrolling functionality and the top middle button acts as a middle button, is that something you have figured out how to do?

schachmat commented:
I hope your exams went well @Lekensteyn, how is the code review going?

Lekensteyn commented:
@schachmat Exams are over, at the moment I am occupied with some power management patches for Linux. I'll try to continue reviewing patches this weekend. I see that the `bit_length()` suggestion is still open, but that is not a blocking issue.

shihrer commented:
I tried out @javitonino's branch and it works on Ubuntu and Fedora.  One thing I noticed is that the smartshift setting persists to another computer.  Is that intended?  I'm not really complaining, it's just unexpected.

lexelby commented:
@shihrer I believe the smart shift setting is stored in the mouse itself.

shihrer commented:
Yeah, that has to be the case.  Just weird that Logitech's software doesn't do it.  I suppose it's to "encourage" people to keep using it.

Lekensteyn commented on lib/solaar/cli/config.py:
```diff
@@ -112,6 +112,12 @@ def run(receivers, args, find_receiver, find_device):
 
+	elif setting.kind == _settings.KIND.range:
+		try:
+			value = int(args.value)
```
This should probably reject values out of range.

Lekensteyn commented on lib/solaar/ui/config_panel.py:
```diff
@@ -76,15 +76,37 @@ def _combo_notify(cbbox, s):
+			self.timer = None
+
+			self.gtk_range = Gtk.Scale()
+			self.gtk_range.set_range(*self.setting.range)
```
Because there is a `@property` `def range()` which returns a tuple, and [`set_range`][1] accepts a `min` and `max` parameter, this works. OK.

 [1]: https://lazka.github.io/pgi-docs/Gtk-3.0/classes/Range.html#Gtk.Range.set_range

Lekensteyn commented on lib/solaar/ui/config_panel.py:
```diff
@@ -76,15 +76,37 @@ def _combo_notify(cbbox, s):
+			if self.gtk_range.get_sensitive():
+				if self.timer:
+					self.timer.cancel()
+				self.timer = _Timer(0.5, lambda: GLib.idle_add(self._write))
```
This could give issues when the user quickly closes the dialog, but I guess that is acceptable.

Lekensteyn commented on lib/logitech_receiver/settings_templates.py:
```diff
@@ -176,6 +179,42 @@ def _feature_smooth_scroll():
+			if threshold == _MAX_SMART_SHIFT_VALUE:
+				threshold = 255
+
+			data = _int2bytes(mode, count=1) + _int2bytes(threshold, count=1) * 2
```
(This sets both the current and (afer reset/default) setting for when the ratchet automatically disengages.)

Lekensteyn commented:
Thank you @javitonino for the PR and others for testing and feedback!

arigit commented:
@javitonino now that smartshift has been working perfectly for a while and was merged - wondering if thumbwheel sensitivity could be added at least via CLI to round up support for this mouse. 
Sorry to be a nag, this is a feature I use while working on windows to help prevent accidental tab switching on firefox, and I miss it in linux

javitonino commented:
@Lekensteyn, thanks for merging. Good to see this in master. If you run into some problems with this or related code, don't hesitate to ping me in an issue/PR so I can help.

@arigit. I think thumbwheel sensitivity is not a mouse-level setting (or at least I have not managed to identify anything from the Windows driver), having to configure it an OS/application level. Using `xset` you can set a divider for the scroll speed. And in Firefox's `about:config` you have `mousewheel.default.delta_multiplier_x` (although I don't know if that applies to tab switching).

arigit commented:
thanks @javitonino - playing with "xinput" I could find that the Horizontal scroll wheel produces button-envets:  buttons 6 and 7 for horizontal scroll.
What I can't find is how to set a "divider" for button presses so that I can make two "horizontal wheel kepresses" count as one.

Otra pointers: 
* xinput --list
* xinput test 11
* xinput list-props 11
* xinput --set-prop 11 'libinput Horizonal Scroll Enabled' 0  (or 1)

Any ideas - anyone?

ps. the firefox 'multiplier' setting does not seem to help

javitonino commented:
@arigit Try this for a divider (by 2, in the horizontal axis):
`xinput set-prop 11 "Evdev Scrolling Distance" 1 2 1`

arigit commented:
Thanks Javi, this is what I get:

`xinput set-prop 11 "Evdev Scrolling Distance" 1 2 1
property 'Evdev Scrolling Distance' doesn't exist, you need to specify its type and format
`

Listing the properties this one doesn't show up,

`xinput list-props 11
Device 'Logitech MX Master':
	Device Enabled (133):	1
	Coordinate Transformation Matrix (135):	1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
	libinput Accel Speed (269):	-0.182511
	libinput Accel Speed Default (270):	0.000000
	libinput Accel Profiles Available (271):	1, 1
	libinput Accel Profile Enabled (272):	1, 0
	libinput Accel Profile Enabled Default (273):	1, 0
	libinput Natural Scrolling Enabled (274):	0
	libinput Natural Scrolling Enabled Default (275):	0
	libinput Send Events Modes Available (253):	1, 0
	libinput Send Events Mode Enabled (254):	0, 0
	libinput Send Events Mode Enabled Default (255):	0, 0
	libinput Left Handed Enabled (276):	0
	libinput Left Handed Enabled Default (277):	0
	libinput Scroll Methods Available (278):	0, 0, 1
	libinput Scroll Method Enabled (279):	0, 0, 0
	libinput Scroll Method Enabled Default (280):	0, 0, 0
	libinput Button Scrolling Button (281):	2
	libinput Button Scrolling Button Default (282):	274
	libinput Middle Emulation Enabled (283):	0
	libinput Middle Emulation Enabled Default (284):	0
	Device Node (256):	"/dev/input/event6"
	Device Product ID (257):	1133, 16449
	libinput Drag Lock Buttons (285):	<no items>
	libinput Horizonal Scroll Enabled (258):	1
$
`

By toggling "libinput Horizontal Scroll Enabled" to 0 or 1 I can disable/enable the thumbwheel but that is as far as I could get. Do you see different properties? I am using Fedora 24 with:

`xinput --version
xinput version 1.6.2
XI version on server: 2.3
`

javitonino commented:
Strange :confused:
```
Device 'Logitech MX Master':
        Device Enabled (138):   1
        Coordinate Transformation Matrix (140): 1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        Device Accel Profile (271):     0
        Device Accel Constant Deceleration (272):       1.000000
        Device Accel Adaptive Deceleration (273):       1.000000
        Device Accel Velocity Scaling (274):    10.000000
        Device Product ID (258):        1133, 16449
        Device Node (259):      "/dev/input/event1"
        Evdev Axis Inversion (275):     0, 0
        Evdev Axes Swap (277):  0
        Axis Labels (278):      "Rel X" (148), "Rel Y" (149), "Rel Horiz Wheel" (268), "Rel Dial" (269), "Rel Vert Wheel" (270)
        Button Labels (279):    "Button Left" (141), "Button Middle" (142), "Button Right" (143), "Button Wheel Up" (144), "Button Wheel Down" (145), "Button Horiz Wheel Left" (146), "Button Horiz Wheel Right" (147), "Button Side" (263), "Button Extra" (264), "Button Forward" (265), "Button Back" (266), "Button Task" (267), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261), "Button Unknown" (261)
        Evdev Scrolling Distance (280): 1, 1, 1
        Evdev Middle Button Emulation (281):    0
        Evdev Middle Button Timeout (282):      50
        Evdev Third Button Emulation (283):     0
        Evdev Third Button Emulation Timeout (284):     1000
        Evdev Third Button Emulation Button (285):      3
        Evdev Third Button Emulation Threshold (286):   20
        Evdev Wheel Emulation (287):    0
        Evdev Wheel Emulation Axes (288):       0, 0, 4, 5
        Evdev Wheel Emulation Inertia (289):    10
        Evdev Wheel Emulation Timeout (290):    200
        Evdev Wheel Emulation Button (291):     4
        Evdev Drag Lock Buttons (292):  0
```

```
xinput version 1.6.2
XI version on server: 2.3
```

Running latest Arch Linux, but it's the same xinput version.

It appears you are using `libinput` for the driver, and this appears to be an `evdev` specific property.

arigit commented:
@javitonino : you are right. libinput is the future though as is the driver used in Wayland and modern distros. I asked in the Wayland list & IRC, the devs asked to open a new bug in systemd to get this started:

https://github.com/systemd/systemd/issues/3947

For other people in this thread: in order to add "sensitivity" support for the MX Master thumbwheel when using Wayland, the Wayland team needs to know the number of "events" generated by the thumbwheel during a full rotation. This can be found using the "evemu-record" utility that can output all mouse wheel events to the screen and carefully rotating the thumbwheel 360 degrees (making a small mark in the thumbwhell helps on this). I counted 14 events but this can be tricky so this is my ask to other users to confirm or correct me on this. If you have the chance to test please post your result here or in the above systemd github issue entry.

Thanks
